// @redocly/cli/lib/index.js
#!/usr/bin/env node
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
require("./assert-node-version");
const yargs = require("yargs");
const types_1 = require("./types");
const openapi_core_1 = require("@redocly/openapi-core");
const preview_docs_1 = require("./commands/preview-docs");
const stats_1 = require("./commands/stats");
const split_1 = require("./commands/split");
const join_1 = require("./commands/join");
const push_1 = require("./commands/push");
const lint_1 = require("./commands/lint");
const bundle_1 = require("./commands/bundle");
const login_1 = require("./commands/login");
const build_docs_1 = require("./commands/build-docs");
const update_version_notifier_1 = require("./update-version-notifier");
const wrapper_1 = require("./wrapper");
const update_version_notifier_2 = require("./update-version-notifier");
if (!('replaceAll' in String.prototype)) {
    require('core-js/actual/string/replace-all');
}
update_version_notifier_1.cacheLatestVersion();
yargs
    .version('version', 'Show version number.', update_version_notifier_2.version)
    .help('help', 'Show help.')
    .parserConfiguration({ 'greedy-arrays': false, 'camel-case-expansion': false })
    .command('stats [api]', 'Show statistics for an API description.', (yargs) => yargs.positional('api', { type: 'string' }).option({
    config: { description: 'Path to the config file.', type: 'string' },
    format: {
        description: 'Use a specific output format.',
        choices: ['stylish', 'json'],
        default: 'stylish',
    },
}), (argv) => {
    process.env.REDOCLY_CLI_COMMAND = 'stats';
    wrapper_1.commandWrapper(stats_1.handleStats)(argv);
})
    .command('split [api]', 'Split an API description into a multi-file structure.', (yargs) => yargs
    .positional('api', {
    description: 'API description file that you want to split',
    type: 'string',
})
    .option({
    outDir: {
        description: 'Output directory where files will be saved.',
        required: true,
        type: 'string',
    },
    separator: {
        description: 'File path separator used while splitting.',
        required: false,
        type: 'string',
        default: '_',
    },
    config: {
        description: 'Path to the config file.',
        requiresArg: true,
        type: 'string',
    },
})
    .demandOption('api'), (argv) => {
    process.env.REDOCLY_CLI_COMMAND = 'split';
    wrapper_1.commandWrapper(split_1.handleSplit)(argv);
})
    .command('join [apis...]', 'Join definitions [experimental].', (yargs) => yargs
    .positional('apis', {
    array: true,
    type: 'string',
    demandOption: true,
})
    .option({
    lint: { description: 'Lint definitions', type: 'boolean', default: false },
    decorate: { description: 'Run decorators', type: 'boolean', default: false },
    preprocess: { description: 'Run preprocessors', type: 'boolean', default: false },
    'prefix-tags-with-info-prop': {
        description: 'Prefix tags with property value from info object.',
        requiresArg: true,
        type: 'string',
    },
    'prefix-tags-with-filename': {
        description: 'Prefix tags with property value from file name.',
        type: 'boolean',
        default: false,
    },
    'prefix-components-with-info-prop': {
        description: 'Prefix components with property value from info object.',
        requiresArg: true,
        type: 'string',
    },
    'without-x-tag-groups': {
        description: 'Skip automated x-tagGroups creation',
        type: 'boolean',
    },
    output: {
        describe: 'Output file',
        alias: 'o',
        type: 'string',
        default: 'openapi.yaml',
    },
    config: {
        description: 'Path to the config file.',
        requiresArg: true,
        type: 'string',
    },
}), (argv) => {
    process.env.REDOCLY_CLI_COMMAND = 'join';
    wrapper_1.commandWrapper(join_1.handleJoin)(argv);
})
    .command('push [api] [maybeDestination] [maybeBranchName]', 'Push an API description to the Redocly API registry.', (yargs) => yargs
    .usage('push [api]')
    .positional('api', { type: 'string' })
    .positional('maybeDestination', { type: 'string' })
    .hide('maybeDestination')
    .hide('maybeBranchName')
    .option({
    organization: {
        description: 'Name of the organization to push to.',
        type: 'string',
        alias: 'o',
    },
    destination: {
        description: 'API name and version in the format `name@version`.',
        type: 'string',
        alias: 'd',
    },
    branch: {
        description: 'Branch name to push to.',
        type: 'string',
        alias: 'b',
    },
    upsert: {
        description: "Create the specified API version if it doesn't exist, update if it does exist.",
        type: 'boolean',
        alias: 'u',
    },
    'batch-id': {
        description: 'Specifies the ID of the CI job that the current push will be associated with.',
        type: 'string',
        requiresArg: true,
        deprecated: true,
        hidden: true,
    },
    'job-id': {
        description: 'ID of the CI job that the current push will be associated with.',
        type: 'string',
        requiresArg: true,
    },
    'batch-size': {
        description: 'Number of CI pushes to expect in a batch.',
        type: 'number',
        requiresArg: true,
    },
    region: { description: 'Specify a region.', alias: 'r', choices: types_1.regionChoices },
    'skip-decorator': {
        description: 'Ignore certain decorators.',
        array: true,
        type: 'string',
    },
    public: {
        description: 'Make the API description available to the public',
        type: 'boolean',
    },
    files: {
        description: 'List of other folders and files to upload',
        array: true,
        type: 'string',
    },
})
    .deprecateOption('batch-id', 'use --job-id')
    .deprecateOption('maybeDestination')
    .implies('job-id', 'batch-size')
    .implies('batch-id', 'batch-size')
    .implies('batch-size', 'job-id'), (argv) => {
    process.env.REDOCLY_CLI_COMMAND = 'push';
    wrapper_1.commandWrapper(push_1.transformPush(push_1.handlePush))(argv);
})
    .command('lint [apis...]', 'Lint definition.', (yargs) => yargs.positional('apis', { array: true, type: 'string', demandOption: true }).option({
    format: {
        description: 'Use a specific output format.',
        choices: [
            'stylish',
            'codeframe',
            'json',
            'checkstyle',
            'codeclimate',
            'summary',
        ],
        default: 'codeframe',
    },
    'max-problems': {
        requiresArg: true,
        description: 'Reduce output to a maximum of N problems.',
        type: 'number',
        default: 100,
    },
    'generate-ignore-file': {
        description: 'Generate an ignore file.',
        type: 'boolean',
    },
    'skip-rule': {
        description: 'Ignore certain rules.',
        array: true,
        type: 'string',
    },
    'skip-preprocessor': {
        description: 'Ignore certain preprocessors.',
        array: true,
        type: 'string',
    },
    'lint-config': {
        description: 'Severity level for config file linting.',
        choices: ['warn', 'error', 'off'],
        default: 'warn',
    },
    config: {
        description: 'Path to the config file.',
        requiresArg: true,
        type: 'string',
    },
    extends: {
        description: 'Override extends configurations (defaults or config file settings).',
        requiresArg: true,
        array: true,
        type: 'string',
    },
}), (argv) => {
    process.env.REDOCLY_CLI_COMMAND = 'lint';
    wrapper_1.commandWrapper(lint_1.handleLint)(argv);
})
    .command('bundle [apis...]', 'Bundle definition.', (yargs) => yargs.positional('apis', { array: true, type: 'string', demandOption: true }).options({
    output: { type: 'string', alias: 'o' },
    format: {
        description: 'Use a specific output format.',
        choices: ['stylish', 'codeframe', 'json', 'checkstyle'],
        default: 'codeframe',
    },
    'max-problems': {
        requiresArg: true,
        description: 'Reduce output to a maximum of N problems.',
        type: 'number',
        default: 100,
    },
    ext: {
        description: 'Bundle file extension.',
        requiresArg: true,
        choices: types_1.outputExtensions,
    },
    'skip-rule': {
        description: 'Ignore certain rules.',
        array: true,
        type: 'string',
    },
    'skip-preprocessor': {
        description: 'Ignore certain preprocessors.',
        array: true,
        type: 'string',
    },
    'skip-decorator': {
        description: 'Ignore certain decorators.',
        array: true,
        type: 'string',
    },
    dereferenced: {
        alias: 'd',
        type: 'boolean',
        description: 'Produce a fully dereferenced bundle.',
    },
    force: {
        alias: 'f',
        type: 'boolean',
        description: 'Produce bundle output even when errors occur.',
    },
    config: {
        description: 'Path to the config file.',
        type: 'string',
    },
    lint: {
        description: 'Lint API descriptions',
        type: 'boolean',
        default: false,
    },
    metafile: {
        description: 'Produce metadata about the bundle',
        type: 'string',
    },
    extends: {
        description: 'Override extends configurations (defaults or config file settings).',
        requiresArg: true,
        array: true,
        type: 'string',
    },
    'remove-unused-components': {
        description: 'Remove unused components.',
        type: 'boolean',
        default: false,
    },
    'keep-url-references': {
        description: 'Keep absolute url references.',
        type: 'boolean',
        alias: 'k',
    },
}), (argv) => {
    process.env.REDOCLY_CLI_COMMAND = 'bundle';
    wrapper_1.commandWrapper(bundle_1.handleBundle)(argv);
})
    .command('login', 'Login to the Redocly API registry with an access token.', (yargs) => __awaiter(void 0, void 0, void 0, function* () {
    return yargs.options({
        verbose: {
            description: 'Include additional output.',
            type: 'boolean',
        },
        region: {
            description: 'Specify a region.',
            alias: 'r',
            choices: types_1.regionChoices,
        },
        config: {
            description: 'Path to the config file.',
            requiresArg: true,
            type: 'string',
        },
    });
}), (argv) => {
    process.env.REDOCLY_CLI_COMMAND = 'login';
    wrapper_1.commandWrapper(login_1.handleLogin)(argv);
})
    .command('logout', 'Clear your stored credentials for the Redocly API registry.', (yargs) => yargs, (argv) => __awaiter(void 0, void 0, void 0, function* () {
    process.env.REDOCLY_CLI_COMMAND = 'logout';
    yield wrapper_1.commandWrapper(() => __awaiter(void 0, void 0, void 0, function* () {
        const client = new openapi_core_1.RedoclyClient();
        client.logout();
        process.stdout.write('Logged out from the Redocly account. ✋\n');
    }))(argv);
}))
    .command('preview-docs [api]', 'Preview API reference docs for the specified definition.', (yargs) => yargs.positional('api', { type: 'string' }).options({
    port: {
        alias: 'p',
        type: 'number',
        default: 8080,
        description: 'Preview port.',
    },
    host: {
        alias: 'h',
        type: 'string',
        default: '127.0.0.1',
        description: 'Preview host.',
    },
    'skip-preprocessor': {
        description: 'Ignore certain preprocessors.',
        array: true,
        type: 'string',
    },
    'skip-decorator': {
        description: 'Ignore certain decorators.',
        array: true,
        type: 'string',
    },
    'use-community-edition': {
        description: 'Use Redoc CE for documentation preview.',
        type: 'boolean',
    },
    force: {
        alias: 'f',
        type: 'boolean',
        description: 'Produce bundle output even when errors occur.',
    },
    config: {
        description: 'Path to the config file.',
        type: 'string',
    },
}), (argv) => {
    process.env.REDOCLY_CLI_COMMAND = 'preview-docs';
    wrapper_1.commandWrapper(preview_docs_1.previewDocs)(argv);
})
    .command('build-docs [api]', 'Produce API documentation as an HTML file', (yargs) => yargs
    .positional('api', { type: 'string' })
    .options({
    o: {
        describe: 'Output destination file.',
        alias: 'output',
        type: 'string',
        default: 'redoc-static.html',
    },
    title: {
        describe: 'Page title.',
        type: 'string',
    },
    disableGoogleFont: {
        describe: 'Disable Google fonts.',
        type: 'boolean',
        default: false,
    },
    t: {
        alias: 'template',
        describe: 'Path to handlebars page template, see https://git.io/vh8fP for the example.',
        type: 'string',
    },
    templateOptions: {
        describe: 'Additional options to pass to the template. Use dot notation, e.g. templateOptions.metaDescription',
    },
    theme: {
        describe: 'Redoc theme.openapi configuration. Use dot notation, e.g. theme.openapi.nativeScrollbars',
    },
    config: {
        describe: 'Path to the config file.',
        type: 'string',
    },
})
    .check((argv) => {
    var _a;
    if (argv.theme && !((_a = argv.theme) === null || _a === void 0 ? void 0 : _a.openapi))
        throw Error('Invalid option: theme.openapi not set');
    return true;
}), (argv) => __awaiter(void 0, void 0, void 0, function* () {
    process.env.REDOCLY_CLI_COMMAND = 'build-docs';
    wrapper_1.commandWrapper(build_docs_1.handlerBuildCommand)(argv);
}))
    .completion('completion', 'Generate completion script.')
    .demandCommand(1)
    .middleware([update_version_notifier_1.notifyUpdateCliVersion])
    .strict().argv;
